File: app/src/main/java/com/maliotis/petros/stormy/weather/Hour.java
package com.maliotis.petros.stormy.weather;

import android.os.Parcel;
import android.os.Parcelable;

import java.text.SimpleDateFormat;
import java.util.Date;

public class Hour implements Parcelable{

    private long mTime;
    private  String mSummary;
    private double mTemperature;
    private String mICon;
    private String mTimezone;

    public long getTime() {
        return mTime;
    }

    public void setTime(long time) {
        mTime = time;
    }

    public String getSummary() {
        return mSummary;
    }

    public void setSummary(String summary) {
        mSummary = summary;
    }

    public int getTemperature() {
        return (int) Math.round(mTemperature);
    }

    public void setTemperature(double temperature) {
        mTemperature = temperature;
        mTemperature = (mTemperature-32)/1.8;
    }

    public String getICon() {
        return mICon;
    }

    public int getIconId(){
        return Forecast.getIconId(mICon);
    }

    public void setICon(String ICon) {
        mICon = ICon;
    }

    public String getTimezone() {
        return mTimezone;
    }

    public void setTimezone(String timezone) {
        mTimezone = timezone;
    }

    public String getHour(){
        SimpleDateFormat formatter = new SimpleDateFormat("H:mm");
        Date date = new Date(mTime * 1000);
        return formatter.format(date);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(mTime);
        dest.writeString(mSummary);
        dest.writeDouble(mTemperature);
        dest.writeString(mICon);
        dest.writeString(mTimezone);
    }

    private  Hour (Parcel in){
        mTime = in.readLong();
        mSummary = in.readString();
        mTemperature = in.readDouble();
        mICon = in.readString();
        mTimezone = in.readString();
    }

    public Hour(){

    }

    public static final Creator<Hour> CREATOR = new Creator<Hour>() {
        @Override
        public Hour createFromParcel(Parcel source) {
            return new Hour(source);
        }

        @Override
        public Hour[] newArray(int size) {
            return new Hour[size];
        }
    };
}

File: app/src/main/java/com/maliotis/petros/weather/weather/Hour.kt
package com.maliotis.petros.weather.weather

import android.os.Parcel
import android.os.Parcelable
import java.text.SimpleDateFormat
import java.util.*

class Hour : Parcelable, Comparable<Hour?> {
    var time: Long = 0
    var summary: String? = null
    private var mTemperature = 0.0
    var iCon: String? = null
    var timezone: String? = null

    val temperature: Int
        get() = Math.round(mTemperature).toInt()

    fun setTemperature(temperature: Double) {
        mTemperature = temperature
        mTemperature = (mTemperature - 32) / 1.8
    }

    val iconId: Int
        get() = Forecast.getIconId(iCon)

    val hour: String
        get() {
            val formatter = SimpleDateFormat("H:mm")
            val date = Date(time * 1000)
            return formatter.format(date)
        }

    val dayOfTheWeekH: String
        get() {
            val formatter = SimpleDateFormat("EEEE")
            formatter.timeZone = TimeZone.getTimeZone(timezone)
            val dateTime = Date(time * 1000)
            return formatter.format(dateTime)
        }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(time)
        dest.writeString(summary)
        dest.writeDouble(mTemperature)
        dest.writeString(iCon)
        dest.writeString(timezone)
    }

    private constructor(`in`: Parcel) {
        time = `in`.readLong()
        summary = `in`.readString()
        mTemperature = `in`.readDouble()
        iCon = `in`.readString()
        timezone = `in`.readString()
    }

    constructor() {}

    override fun compareTo(other: Hour?): Int {
        return if (other != null)
            this.hour.compareTo(other.hour)
        else 0
    }

    companion object {
        val CREATOR: Parcelable.Creator<Hour> = object : Parcelable.Creator<Hour> {
            override fun createFromParcel(source: Parcel): Hour? {
                return Hour(source)
            }

            override fun newArray(size: Int): Array<Hour?> {
                return arrayOfNulls(size)
            }
        }
    }
}